### PR TITLE
feat: Feature/adding remote online store

### PR DESCRIPTION
**Pull request by lokeshrangineni**
# What this PR does / why we need it:
Adds the new online store of type "remote" which will interact with remote feature server. Able to read and write the features only at the moment. 

# Which issue(s) this PR fixes:
https://github.com/feast-dev/feast/issues/4121

### CODE REVIEW

**Comment by lokeshrangineni** [comment]
cc: @tokoko @dmartinol

File: sdk/python/tests/integration/feature_repos/universal/online_store/remote.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from tests.integration.feature_repos.universal.online_store_creator import (
+    OnlineStoreCreator,
+)
+
+class RemoteOnlineStoreCreator(OnlineStoreCreator):
+    _FEAST_SERVER_TIMEOUT_SEC:int = 30
+    def __init__(self, project_name: str, **kwargs):
+        super().__init__(project_name)
+        
+        for key, value in kwargs.items():
```
Comment (tokoko) [inline sdk/python/tests/integration/feature_repos/universal/online_store/remote.py:18]: Not a deal-breaker, but we should probably change `OnlineStoreCreator` similar to `DataSourceCreator`, meaning add a `setup` method that's called with a registry config.

File: sdk/python/tests/integration/feature_repos/repo_configuration.py [hunk 2]
```diff
@@ -153,6 +160,7 @@
     AVAILABLE_ONLINE_STORES["datastore"] = ("datastore", None)
     AVAILABLE_ONLINE_STORES["snowflake"] = (SNOWFLAKE_CONFIG, None)
     AVAILABLE_ONLINE_STORES["bigtable"] = (BIGTABLE_CONFIG, None)
+    AVAILABLE_ONLINE_STORES["remote"] = (REMOTE_ONLINE_CONFIG, None)
```
Comment (tokoko) [inline sdk/python/tests/integration/feature_repos/repo_configuration.py:163]: Is this necessary? `remote` should already be in the list from above

File: sdk/python/feast/infra/online_stores/remote.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                "push_source_name": table.name,
+                "df": json_df,
+                "to": "online"
+            }
+        
+        return req_body
+
+    def _check_if_feature_requested(self, feature_name, requested_features):
+        for requested_feature in requested_features:
+            if feature_name in requested_feature:
```
Comment (tokoko) [inline sdk/python/feast/infra/online_stores/remote.py:150]: Isn't this a substring check? Shouldn't a simple equality check be sufficient? Or maybe there's an edge case because of `full_feature_names`?

File: sdk/python/feast/infra/online_stores/remote.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        headers = {'Content-type': 'application/json', 'Accept': 'text/plain'}     
+        response = requests.post(f'{config.online_store.path}/push', data=json.dumps(req_body), headers=headers)
+        if response.status_code == 200:
+            logger.debug("Able to write the online features from the feature server.")
+        else:
+            error_message = f"Unable to write the data to online store using feature server API. Error_code={response.status_code}, error_message={response.reason}"
+            logger.error(error_message)
+            raise RuntimeError(error_message)
+
+    def online_read(
```
Comment (tokoko) [inline sdk/python/feast/infra/online_stores/remote.py:72]: I'm realizing we will probably have the same problem with `FeatureService`s here as in offline server discussed [here](https://github.com/RHEcosystemAppEng/feast/pull/3#issuecomment-2127894618). cc @dmartinol 

This is also probably a very chatty protocol with a list of features always transmitted on the wire even if feature service was requested.

File: sdk/python/feast/infra/online_stores/remote.py (hunk 3 above)
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py:150]: Yes, that is correct. Just to incorporate when full_feature_names is used. Please let me know if there is any better way.

File: sdk/python/feast/infra/online_stores/remote.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+from feast.protos.feast.types.Value_pb2 import Value as ValueProto
+from feast.repo_config import FeastConfigBaseModel
+from feast.type_map import python_values_to_proto_values
+from feast.value_type import ValueType
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteOnlineStoreConfig(FeastConfigBaseModel):
+    """Online store config for Redis store"""
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:34]: Redis?

File: sdk/python/feast/infra/online_stores/remote.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    def online_write_batch(
+        self,
+        config: RepoConfig,
+        table: FeatureView,
+        data: List[
+            Tuple[EntityKeyProto, Dict[str, ValueProto], datetime, Optional[datetime]]
+        ],
+        progress: Optional[Callable[[int], Any]],
+    ) -> None:
+        pass
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:58]: Is this a `pass` or `NotImplementedError`?

File: sdk/python/feast/infra/online_stores/remote.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        if response.status_code == 200:
+            logger.debug("Able to retrieve the online features from feature server.")
+            response_json = json.loads(response.text)
+            event_ts = self._get_event_ts(response_json)
+            # Iterating over results and converting the API results in column format to row format.
+            feature_value_index = 0
+            result_tuples: List[
+                Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]
+            ] = []
+            while feature_value_index < len(entity_keys):
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:85]: `for feature_value_index in range(len(entity_keys))` ?

File: sdk/python/feast/infra/online_stores/remote.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+                feature_values_dict = dict()
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:86]: shouldn't you also specify the type?

File: sdk/python/feast/infra/online_stores/remote.py (hunk 3 above)
Comment (dmartinol) [reply sdk/python/feast/infra/online_stores/remote.py:151]: this returns True if we feature name is a substring of the requested feature, it can return false positives IMO.

File: sdk/python/feast/infra/online_stores/remote.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        for requested_feature in requested_features:
+            if feature_name in requested_feature:
+                return True
+        return False
+
+    def _get_event_ts(self, response_json) -> datetime:
+        event_ts = ""
+        for index, result in enumerate(response_json["results"]):
+            if index == 1:
+                event_ts = result["event_timestamps"][0]
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:159]: isn't this equivalent to the following?
```py
# maybe check len of first array
event_ts = response_json["results"][1]["event_timestamp"][0]
```

File: sdk/python/tests/integration/feature_repos/repo_configuration.py [hunk 10]
```diff
@@ -509,6 +498,13 @@ def construct_test_environment(
             cache_ttl_seconds=1,
         )
 
+    if test_repo_config.online_store_creator:
```
Comment (dmartinol) [inline sdk/python/tests/integration/feature_repos/repo_configuration.py:501]: why did you move it down?

File: sdk/python/feast/infra/online_stores/remote.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+            ] = []
+            while feature_value_index < len(entity_keys):
+                feature_values_dict = dict()
+                for index, feature_name in enumerate(
+                    response_json["metadata"]["feature_names"]
+                ):
+                    if self._check_if_feature_requested(
+                        feature_name, requested_features
+                    ):
+                        is_present = (
```
[the remark this answers is not in the document]
Comment (dmartinol) [reply sdk/python/feast/infra/online_stores/remote.py:93]: and you probably don't even need the `is_present` variable, but 🤷‍♂️

File: sdk/python/feast/infra/online_stores/remote.py (hunk 9 above)
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py:159]: just trying to avoid the array index exception, i can check the length of the array before instead of iterating over it.

File: sdk/python/feast/infra/online_stores/remote.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py:93]: this logic is evolved and i didn't realize that it is not needed anymore. i have removed it. thanks for pointing it out.

File: sdk/python/tests/integration/feature_repos/repo_configuration.py (hunk 10 above)
Comment (lokeshrangineni) [reply sdk/python/tests/integration/feature_repos/repo_configuration.py:501]: it is not needed to move now, initially made some changes to incorporate that logic into the creator framework at that time there was some dependency but later that plan was dropped in my case.

File: sdk/python/feast/infra/online_stores/remote.py (hunk 6 above)
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py:58]: i don't want to through the hard stop as this life cycle method was used in some of the providers so used pass instead of throwing exception. please let me know in case if you think it should be other way.

Comment (dmartinol) [reply sdk/python/feast/infra/online_stores/remote.py:58]: @tokoko what’s the preferred approach to ensure application consistency?

File: docs/reference/online-stores/remote.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+The registry is pointing to registry of remote feature store. If it is not accessible then should be configured to use remote registry.
+
+{% code title="feature_store.yaml" %}
+```yaml
+project: my-local-project
+  registry: /remote/data/registry.db
+  provider: local
+  online_store:
+    path: http://localhost:6566
```
Comment (dmartinol) [inline docs/reference/online-stores/remote.md:17]: shouldn't we have host and port fields instead? this is how we configured the remote offline:
```
offline_store:
    type: remote
    host: localhost
    port: 8815
```

Comment (lokeshrangineni) [reply docs/reference/online-stores/remote.md:17]: I have noticed this for offline_Store. We are already having path field for online store so reusing the same instead of introducing new properties. Remote registry is also doing the same. Also original recommendation by @tokoko but happy to change the code to keep it consistent across. 

https://github.com/feast-dev/feast/blob/master/sdk/python/feast/infra/registry/remote.py#L30

File: sdk/python/feast/infra/online_stores/remote.py (hunk 3 above)
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py:147]: updated the code after performing some tests, it makes sense to use ==.

File: sdk/python/feast/infra/online_stores/remote.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            result_tuples: List[
+                Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]
+            ] = []
+            for feature_value_index in range(len(entity_keys)):
+                feature_values_dict: Dict[str, ValueProto] = dict()
+                for index, feature_name in enumerate(
+                    response_json["metadata"]["feature_names"]
+                ):
+                    if (
+                        requested_features is not None
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:90]: what if requested_features is None, instead? should we return all features or no features? (in which case, why did we call the remote API? 🤔 )

File: sdk/python/feast/infra/online_stores/remote.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+        req_body = json.dumps(
+            {
+                "features": api_requested_features,
+                "entities": {entity_key: entity_values},
+            }
+        )
+        return req_body
+
+    def _check_if_feature_requested(self, feature_name, requested_features):
```
Comment (dmartinol) [inline sdk/python/feast/infra/online_stores/remote.py:145]: Is this `_check_if_feature_requested` function still in use?

File: docs/reference/online-stores/remote.md (hunk 12 above)
Comment (tokoko) [reply docs/reference/online-stores/remote.md:17]: I also simply avoided introducing new properties. Sounds more convenient as a single url to me, you always have some space to play with protocols and so on...

File: sdk/python/feast/infra/online_stores/remote.py (hunk 13 above)
Comment (tokoko) [reply sdk/python/feast/infra/online_stores/remote.py:90]: None means all features

File: sdk/python/tests/integration/online_store/test_remote_online_store.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
+import os
+import tempfile
+from datetime import datetime
+from multiprocessing import Process
```
Comment (dmartinol) [inline sdk/python/tests/integration/online_store/test_remote_online_store.py:4]: should we use Subprocess as it was recently done [here](https://github.com/feast-dev/feast/pull/4262/commits/dec05c91adf9ed7415961ad6bc90ee91b10227b3#diff-93e66e2dc557553d09ce1e0fc77b45003e1edcc8592979886970d5a878f731b2R398)?

File: sdk/python/feast/infra/online_stores/remote.py (hunk 13 above)
Comment (dmartinol) [reply sdk/python/feast/infra/online_stores/remote.py:90]: so we should probably change to
```py
if requested_features is None or feature_name in requested_features:
```

Comment (tokoko) [reply sdk/python/feast/infra/online_stores/remote.py:90]: That feels like something that should have been part of integration tests already 😆

Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py:90]: it is already part of the integration test case, just didn't realize until exploring more details about full_feature_names flag.

File: sdk/python/tests/conftest.py [hunk 16]
```diff
@@ -173,7 +173,15 @@ def simple_dataset_2() -> pd.DataFrame:
 
 def start_test_local_server(repo_path: str, port: int):
     fs = FeatureStore(repo_path)
-    fs.serve("localhost", port, no_access_log=True)
+    fs.serve(
```
[the remark this answers is not in the document]
Comment (lokeshrangineni) [reply sdk/python/tests/conftest.py:176]: Yes, those are not defined as optional for feature_store.serve() method.

File: sdk/python/feast/infra/online_stores/remote.py (hunk 14 above)
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py]: removed it.

File: sdk/python/tests/integration/online_store/test_remote_online_store.py (hunk 15 above)
Comment (lokeshrangineni) [reply sdk/python/tests/integration/online_store/test_remote_online_store.py]: Implemented it. please take a look.

File: sdk/python/tests/conftest.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tokoko) [reply sdk/python/tests/conftest.py:176]: Other parameters have [default values](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_store.py#L2514) set. You don't need to provide them, only `host` and `port` are mandatory.

File: sdk/python/tests/integration/online_store/test_remote_online_store.py (hunk 15 above)
Comment (dmartinol) [reply sdk/python/tests/integration/online_store/test_remote_online_store.py]: Lgtm, but please @redhatHameed can you verify it is consistent with the offline implementation?

File: sdk/python/tests/conftest.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (lokeshrangineni) [reply sdk/python/tests/conftest.py:176]: @tokoko - I see that feature_store is modified [recently](https://github.com/feast-dev/feast/commit/fa744380ad48ee394a05b2f600db5fb0a99c03aa#diff-ff4e2a65c8c29b7971edc77a61a7e970d410d77d9f173f72714d1bb49e2509a2R2451) after i started working on this change. The default values have been added to original API so that I can revert this change.

[the remark this answers is not in the document]
Comment (lokeshrangineni) [reply sdk/python/tests/conftest.py]: reverted this change, please take a look.

File: sdk/python/tests/integration/online_store/test_remote_online_store.py (hunk 15 above)
Comment (redhatHameed) [reply sdk/python/tests/integration/online_store/test_remote_online_store.py]: Yup looks lgtm that implementation

File: Makefile [hunk 17]
```diff
@@ -377,6 +377,9 @@ kill-trino-locally:
 install-protoc-dependencies:
 	pip install --ignore-installed protobuf==4.24.0 "grpcio-tools>=1.56.2,<2" mypy-protobuf==3.1.0
 
+install-feast-ci-locally:
+	pip install -e ".[ci]"
+
```
Comment (redhatHameed) [inline Makefile:382]: does we required this 
How it differentiate with ```pip install -e ".[dev]"```

**Review by redhatHameed — COMMENTED** [review]
lgtm minor comments

File: Makefile (hunk 17 above)
Comment (lokeshrangineni) [reply Makefile:382]: just it is a additional target to install dev specific libraries, i had to install this on local to test and resolve one of the issue. 
https://github.com/feast-dev/feast/blob/master/setup.py#L158-L224

Comment (tokoko) [reply Makefile:382]: @lokeshrangineni It's not mandatory of course, but you can also use `make install-python-ci-dependencies-uv` per dev [guide](https://github.com/feast-dev/feast/blob/master/docs/project/development-guide.md#feast-python-sdk--cli). Just a lot better experience imho once you set some stuff up. It also uses pinned requirements files instead of installing latest versions which means your environment is similar to where github runs the tests.

Comment (tokoko) [reply Makefile:382]: P.S. `dev` and `ci` extras are identical now. Just decided to leave both of them not to mess with anyone's workflow

File: sdk/python/feast/infra/online_stores/remote.py (hunk 6 above)
Comment (tokoko) [reply sdk/python/feast/infra/online_stores/remote.py:58]: Sorry, missed this. +1 for `NotImplementedError`. `pass` in this context (without the method marked as abstract) just returns `None` and may send you on very elaborate debugging journeys.

Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py]: updated the PR with `NotImplementedError `

File: Makefile (hunk 17 above)
Comment (lokeshrangineni) [reply Makefile]: if both are doing the same and `make install-python-ci-dependencies-uv` is recommended then better to remove this target. Updated the PR.

File: sdk/python/feast/infra/online_stores/remote.py (hunk 6 above)
Comment (lokeshrangineni) [reply sdk/python/feast/infra/online_stores/remote.py]: It is throwing run time error so had to revert the change for update and teardown methods as it is mandatory. Please review the updated code.

**Review by tokoko — APPROVED** [review]
lgtm

**Comment by lokeshrangineni** [comment]
After [approved](https://github.com/feast-dev/feast/pull/4226#pullrequestreview-2117133596) by @tokoko, just rebased the branch with feast-dev:master branch.  No changes made to PR after the approval.

**Review by franciscojavierarceo — APPROVED** [review]
